Enhancements for issue 8269

Aka2210 opened the pull request:
This PR introduces a new function `directed_steiner_tree` under
`networkx.algorithms.approximation`.

- Implements an approximation algorithm for the **Directed Steiner Tree (DST)** problem.
- Based on Charikar, Chekuri, Cheung, Dai, Goel, Guha, Li (1999):
  *Approximation Algorithms for Directed Steiner Problems*.
- Complements the existing undirected Steiner Tree approximation
  (`nx.algorithms.approximation.steiner_tree`).

dschult commented on networkx/algorithms/approximation/steinertree.py:
```diff
@@ -248,3 +249,179 @@ def steiner_tree(G, terminal_nodes, weight="weight", method=None):
+    G : DiGraph
+        A directed graph.
+    root : node
+        The root node.
```
Can you say very briefly how this is used (could be in a short description of the algorithm -- perhaps in a paragraph after the one-liner description above, and before the Parameters section).

Aka2210 replied:
Sure, I’ve added the extra clarification to the docstring.

dschult commented on networkx/algorithms/approximation/steinertree.py:
```diff
@@ -248,3 +249,179 @@ def steiner_tree(G, terminal_nodes, weight="weight", method=None):
+        A directed graph.
+    root : node
+        The root node.
+    terminals : set
```
Since terminals can be any iterable the way you've got the conversions to `set`, this can be declared as an `iterable of nodes` rather than a set.  right?

dschult commented on networkx/algorithms/approximation/steinertree.py:
```diff
@@ -248,3 +249,179 @@ def steiner_tree(G, terminal_nodes, weight="weight", method=None):
+    return H
+
+
+def directed_steiner_tree(
```
Let's add a decorator to check for undirected inputs (and then remove the lines where the check is manually done here).
`@nx.utils.not_implemented_for("directed")`

And we should think about multigraphs. Does the regular algorithm work without changes for a MultiDiGraph?  If not, we should put a similar decorator with `"multigraph"`.


For the function signature, let's decide whether the keyword args should be "keyword-only". That is, we want people to write out the names of any "long list" of keyword args so that the call is more readable.  I would suggest that `G`, `root` and `terminals` be allowed to be positional parameters, while the rest require the argument name to be included in any call. To implement this suggestion, put `, *` after `terminals` in the signature.

Aka2210 replied:
Fixed in 8119c05d and e28f9ea3, thanks!

dschult commented on networkx/algorithms/approximation/steinertree.py:
```diff
@@ -248,3 +249,179 @@ def steiner_tree(G, terminal_nodes, weight="weight", method=None):
+    covered = set(H.nodes) & set(reachable_terminals)
+    if len(covered) < min_terminals:
+        raise nx.NetworkXUnfeasible(
+            f"Directed Steiner tree could not cover at least {min_terminals} terminals {list(reachable_terminals)}"
```
Just trying to keep the line length down, and state what is being printed.
```suggestion
            "Directed Steiner tree could not cover at least "
            f"{min_terminals}.\nReachable terminals: {list(reachable_terminals)}"
```

Aka2210 replied:
Fixed in 8119c05d, thanks!

dschult commented on networkx/algorithms/approximation/steinertree.py:
```diff
@@ -248,3 +249,179 @@ def steiner_tree(G, terminal_nodes, weight="weight", method=None):
+def _directed_steiner_tree(G, root, terminals, min_terminals, cutoff, weight):
+    """Recursive helper for directed_steiner_tree."""
+    H = nx.DiGraph()
+    if min_terminals <= 0 or not terminals or cutoff == 0:
+        return H
```
This is a private functuion so we shouldn't need to check the inputs. They've already been checked before calling this, right?

dschult reviewed (commented):
Thanks for this!!

We should connect the functions to the docs so they show up. A line in the Steiner Tree section of `doc/reference/algorithms/approximation.rst` should do the trick.

The tests should run just fine. I'd prefer them to call `nx.directed_steiner_tree` rather than importing the function at the top (b/c it tests the import namespace is working; and because tests often get copy/pasted/adapted into code so it's nice to have them self-contained). This is very minor -- and the module's other functions don't do this. So we can make this tweak as a follow-up PR instead.

Doc string for the public function should have the paremeter section by-line for each parameter say what the default is (for optional kwargs).  Something like:
`cutoff : int, optional (default: inf)`

:)

Aka2210 commented on networkx/algorithms/approximation/steinertree.py:
```diff
@@ -248,3 +249,179 @@ def steiner_tree(G, terminal_nodes, weight="weight", method=None):
+    if G.number_of_edges() == 0:
+        return float("inf")
+
+    terminals = set(terminals)
```
Fixed in 8119c05d, thanks!

Aka2210 commented:
All comments have now been addressed:
- Added `directed_steiner_tree` to docs
- Updated docstrings with default values and added note on MultiDiGraph handling
- Enabled support for MultiDiGraph inputs by collapsing parallel edges to the minimum-weight edge (7e92f1a)
- Made `cutoff` and `min_terminals` keyword-only 
- Switched to `@not_implemented_for("undirected")` and updated test to expect `NetworkXNotImplemented`

Please let me know if I missed anything or if further changes are needed.  
Thanks again for the detailed and helpful review!

amcandio commented on networkx/algorithms/approximation/steinertree.py:
```diff
@@ -248,3 +249,206 @@ def steiner_tree(G, terminal_nodes, weight="weight", method=None):
+    while q:
+        node, depth = q.popleft()
+        if depth == cutoff:
+            continue
```
You should be able to break here

Aka2210 replied:
Thanks for pointing that out! This issue has already disappeared after switching to nx.single_source_shortest_path_length.

amcandio commented on networkx/algorithms/approximation/steinertree.py:
```diff
@@ -248,3 +249,206 @@ def steiner_tree(G, terminal_nodes, weight="weight", method=None):
+    return H
+
+
+def _reachable_within(G, source, cutoff):
```
You can reuse nx.single_source_shortest_path_length which has a cutoff parameter and it also does a BFS

Aka2210 replied:
_reachable_within is functionally equivalent to the level-by-level BFS used in nx.single_source_shortest_path_length, but it avoids rebuilding lists at each level so it has slightly smaller constant factors. Both versions work fine — would you prefer consistency with nx.single_source_shortest_path_length, or keeping _reachable_within for efficiency?

amcandio replied:
Unless this part of the code affects performance significantly, I'd reuse preexisting functions. On its current form the nx public method can be faster because it breaks once the cutoff is reached.

amcandio commented on networkx/algorithms/approximation/steinertree.py:
```diff
@@ -248,3 +249,206 @@ def steiner_tree(G, terminal_nodes, weight="weight", method=None):
+
+
+@nx.utils.not_implemented_for("undirected")
+def directed_steiner_tree(
```
But/preference: can we put the public method first?

Aka2210 replied:
Sure, I can reorder so the public method appears first.

amcandio commented on networkx/algorithms/approximation/steinertree.py:
```diff
@@ -248,3 +249,206 @@ def steiner_tree(G, terminal_nodes, weight="weight", method=None):
+        for v in G.successors(root):
+            for n in range(1, len(terminals) + 1):
+                sub_tree = _directed_steiner_tree(
+                    G, v, terminals, n, cutoff - 1, weight
```
You should document that n>1000 can lead to max recursion errors

dschult replied:
This would be cool to make non-recursive -- it is usually lower memory, often faster and avoids the max recursion errors. But we could convert later if you prefer. The idea is that you store the info needed across function calls in a queue instead of calling the function (which does much more than store the info you need).

amcandio commented on networkx/algorithms/approximation/steinertree.py:
```diff
@@ -248,3 +249,206 @@ def steiner_tree(G, terminal_nodes, weight="weight", method=None):
+           directed Steiner problems. *Journal of Algorithms*, 33(1), 73–91.
+           https://doi.org/10.1006/jagm.1999.1042
+    """
+    if G.is_multigraph():
+        G = _collapse_multigraph_to_digraph(G, weight)
```
Can we do this after input validation?

Aka2210 commented on networkx/algorithms/approximation/steinertree.py:
```diff
@@ -248,3 +249,195 @@ def steiner_tree(G, terminal_nodes, weight="weight", method=None):
+                if sub_density < min_density:
+                    min_sub_tree = sub_tree
+                    min_density = sub_density
+        covered_terminals = {n for n in min_sub_tree.nodes if n in terminals}
```
Thanks! But I feel keeping .nodes makes the intent clearer and improves readability. Do you have other considerations in mind?

dschult replied:
To keep the intent clear but shift to using the NetworkX style choice of graph iteration returning nodes, use something like:
```python
        covered_terminals = {node for node in min_sub_tree if node in terminals}
```

Aka2210 commented on networkx/algorithms/approximation/steinertree.py:
```diff
@@ -248,3 +249,195 @@ def steiner_tree(G, terminal_nodes, weight="weight", method=None):
+    return H
+
+
+def _directed_steiner_tree_density(G, terminals, weight):
```
Fixed in ecc83104, thanks!

Aka2210 commented on networkx/algorithms/approximation/steinertree.py:
```diff
@@ -248,3 +249,193 @@ def steiner_tree(G, terminal_nodes, weight="weight", method=None):
+
+    This implementation follows the greedy density-based approach of
+    Charikar et al. (1999). The algorithm repeatedly grows partial trees
+    rooted at the given source until all required terminals are spanned.
```
Thanks, I’ve updated the docstring accordingly.

Aka2210 commented:
Apologies for the earlier confusion.  
After re-reading Sections 2 and 3 of the paper I realized I had overlooked an
important detail:

> “between every pair of vertices u and v, there exists an edge (u,v) of cost
> equal to the shortest path distance from u to v in G.”

This requires constructing the **metric closure** of the input graph and, after
the algorithm finishes, expanding the virtual edges back into their actual
shortest paths.

I also corrected the terminology: what I previously referred to as *cutoff*
should be renamed to **levels**, consistent with the paper. The docstring now
documents:

- Practical choices: `levels=2` or `3` (larger values become inefficient).
- Theoretical guarantee: `levels=log(k)` gives the best `O(log k)` approximation.
- The analysis assumes `i > 1`. Although the code still produces results for
  `levels=1`, this case is outside the scope of the proof.

The docstring (parameters and notes) has been updated, and the implementation
now explicitly performs closure + expansion.

dschult commented:
I think it should match what `steiner_tree` does. If that makes sense.

dschult commented on networkx/algorithms/approximation/steinertree.py:
```diff
@@ -263,3 +264,303 @@ def steiner_tree(G, terminal_nodes, weight="weight", method=None):
+    Approximate solution to the Directed Steiner Tree problem.
+
+    This implementation follows the greedy density-based approach of
+    Charikar et al. (1999). The algorithm repeatedly grows partial trees
```
Let's make this  link to the reference/citation:  `follows the greedy density-based approach in [1]_  Similarly for the other references to this paper in the doc_string.

dschult commented on networkx/algorithms/approximation/steinertree.py:
```diff
@@ -263,3 +264,303 @@ def steiner_tree(G, terminal_nodes, weight="weight", method=None):
+    G_expanded = nx.DiGraph()
+    for u, v, data in G_closure.edges(data=True):
+        path_nodes = data["path"]
+        for a, b in zip(path_nodes[:-1], path_nodes[1:]):
```
```suggestion
        for a, b in nx.utils.pairwise(path_nodes):
```
